src/rhash/hash.c

```diff
@@ -679,6 +679,325 @@ static int rc_hash_7800(char hash[33], const uint8_t* buffer, size_t buffer_size
+      }
+
+    /* Write the pointer and length of the data we record about this file */
+    hashindex->data = hashdata;
+    hashindex->length = filename_len + 1 + 4 + 8;
+    hashindex++;
+
+    /* Convert and store the file name in the hash data buffer */
+    for (name = (cdir + RC_ZIP_CENTRAL_DIR_HEADER_SIZE), name_end = name + filename_len; name != name_end; name++)
+    {
+      *(hashdata++) =
+        (*name == '\\' ? '/' : /* convert back-slashes to regular slashes */
+        (*name >= 'A' && *name <= 'Z') ? (*name | 0x20) : /* convert upper case letters to lower case */
+        *name); /* else use the byte as-is */
+    }
+
+    /* Add zero terminator, CRC32 and decompressed size to the hash data buffer */
+    *(hashdata++) = '\0';
+    RC_ZIP_WRITE_LE32(hashdata, crc);
+    hashdata += 4;
+    RC_ZIP_WRITE_LE64(hashdata, decomp_size);
+    hashdata += 8;
+
+    if (verbose_message_callback)
+    {
+      char message[1024];
+      snprintf(message, sizeof(message), "File in ZIP: %.*s (%u bytes, CRC32 = %08X)", filename_len, (const char*)(cdir + RC_ZIP_CENTRAL_DIR_HEADER_SIZE), (unsigned)decomp_size, crc);
+      verbose_message_callback(message);
+    }
+  }
+
+  if (verbose_message_callback)
+  {
+    char message[1024];
+    snprintf(message, sizeof(message), "Hashing %u files in ZIP archive", (unsigned)(hashindex - hashindices));
+    verbose_message_callback(message);
+  }
+
+  /* Sort the file list indices */
+  qsort(hashindices, (hashindex - hashindices), sizeof(struct rc_hash_zip_idx), rc_hash_zip_idx_sort);
+
+  /* Hash the data in the order of the now sorted indices */
+  for (; hashindices != hashindex; hashindices++)
+    md5_append(md5, hashindices->data, (int)hashindices->length);
+
+  free(alloc_buf);
+  return 1;
+
+  #undef RC_ZIP_READ_LE16
+  #undef RC_ZIP_READ_LE32
+  #undef RC_ZIP_READ_LE64
+  #undef RC_ZIP_WRITE_LE32
+  #undef RC_ZIP_WRITE_LE64
+}
+
+static int rc_hash_ms_dos(char hash[33], const char* path)
+{
+  md5_state_t md5;
+  size_t path_len;
+  int res;
+
+  void* file_handle = rc_file_open(path);
+  if (!file_handle)
+    return rc_hash_error("Could not open file");
+
+  /* hash the main content zip file first */
+  md5_init(&md5);
+  res = rc_hash_zip_file(&md5, file_handle);
+  rc_file_close(file_handle);
+
+  if (!res)
+    return 0;
+
+  /* if this is a .dosz file, check if an associated .dosc file exists */
+  path_len = strlen(path);
+  if (path[path_len-1] == 'z' || path[path_len-1] == 'Z')
+  {
+    char *dosc_path = strdup(path);
+    if (!dosc_path)
+        return rc_hash_error("Could not allocate temporary buffer");
+
+    /* swap the z to c and use the same capitalization, hash the file if it exists*/
+    dosc_path[path_len-1] = (path[path_len-1] == 'z' ? 'c' : 'C');
+    file_handle = rc_file_open(dosc_path);
+    free((void*)dosc_path);
+    if (file_handle)
+    {
+      rc_hash_zip_file(&md5, file_handle);
```

Comment:
failure return should be captured and returned (`res=`, `if !res`)

Reply:
My first thought was to basically ignore a DOSC that exists but is not a ZIP (basically like if it does not exist at all) but it makes sense to fail because `rc_hash_zip_file` will already have logged errors about it. And really the .dosc should be a proper ZIP file. So the latest commit fixed this.